x/sockopt/sockopt.go

```diff
@@ -50,15 +60,51 @@ var _ HasHopLimit = (*hopLimitOption)(nil)
 
 // TCPOptions represents options for TCP connections.
 type TCPOptions interface {
+	HasWaitUntilBytesAreSent
 	HasHopLimit
 }
 
 type tcpOptions struct {
 	hopLimitOption
+
+	conn *net.TCPConn
+
+	// Timeout after which we return an error
+	waitingTimeout time.Duration
+	// Delay between checking the socket
+	waitingDelay time.Duration
 }
 
 var _ TCPOptions = (*tcpOptions)(nil)
 
+func (o *tcpOptions) SetWaitingTimeout(timeout time.Duration) {
+	o.waitingTimeout = timeout
+}
+
+func (o *tcpOptions) SetWaitingDelay(delay time.Duration) {
+	o.waitingDelay = delay
+}
+
+func (o *tcpOptions) OsSupportsWaitingUntilBytesAreSent() bool {
+	return isConnectionSendingBytesImplemented()
+}
+
+func (o *tcpOptions) WaitUntilBytesAreSent() error {
```

Comment:
This functionality is too high-level to live here.

I'd say we just need the bytes not sent and state options.

Then you can implement the logic to wait completely separate.